File: python/spiral_matrix.py
def spiral_matrix(size):
    mat = [[0] * size for _ in range(size)]
    n = 1
    x, y = 0, 0
    dirs = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    d = 0
    dx, dy = dirs[d]
    while n <= size * size:
        mat[x][y] = n
        n += 1
        nx = x + dx
        ny = y + dy
        if nx < 0 or ny < 0 or nx >= size or ny >= size or mat[nx][ny] != 0:
            d = (d + 1) % 4
        dx, dy = dirs[d]
        x += dx
        y += dy
    return mat

File: go/spiral_matrix.go
package spiralmatrix

func SpiralMatrix(size int) [][]int {
	mat := make([][]int, size)
	for i := 0; i < size; i++ {
		mat[i] = make([]int, size)
	}

	dx := 0
	dy := 1
	x := 0
	y := 0

	for i := 1; i <= size*size; i++ {
		mat[x][y] = i
		nx := x + dx
		ny := y + dy
		if nx < 0 || nx >= size || ny < 0 || ny >= size || mat[nx][ny] > 0 {
			t := dx
			dx = dy
			dy = -t
		}
		x += dx
		y += dy
	}

	return mat
}
